finality-aleph/src/validator_network/service.rs

```diff
@@ -150,76 +154,122 @@ impl<D: Data, A: Data + Debug, ND: Dialer<A>, NL: Listener> Service<D, A, ND, NL
+                            },
+                            false => match self.manager.send_to(&peer_id, data) {
+                                Ok(_) => trace!(target: "validator-network", "Sending data to {}.", peer_id),
+                                Err(e) => trace!(target: "validator-network", "Failed sending to {}: {}", peer_id, e),
+                            },
                         }
                     },
                 },
-                // received tuple (peer_id, exit_handle) from a spawned worker
-                // that has just established an incoming connection
-                // pass the tuple to the manager to register the connection
-                // the manager will be responsible for killing the worker if necessary
-                Some((peer_id, Some(exit), _)) = incoming_workers.next() => {
-                    use AddResult::*;
-                    match self.manager.add_incoming(peer_id.clone(), exit) {
-                        Uninterested => info!(target: "validator-network", "Peer {} connected to us despite out lack of interest.", peer_id),
-                        Added => info!(target: "validator-network", "New incoming connection for peer {}.", peer_id),
-                        Replaced => info!(target: "validator-network", "Replaced incoming connection for peer {}.", peer_id),
-                    }
-                },
-                // received information from a spawned worker managing an outgoing connection
+                // received information from a spawned worker managing a connection
                 // check if we still want to be connected to the peer, and if so, spawn a new worker or actually add proper connection
-                Some((peer_id, maybe_data_for_network, _)) = outgoing_workers.next() => {
-                    use AddResult::*;
-                    if let Some(addresses) = self.manager.peer_addresses(&peer_id) {
-                        match maybe_data_for_network {
-                            Some(data_for_network) => match self.manager.add_outgoing(peer_id.clone(), data_for_network) {
-                                Uninterested => warn!(target: "validator-network", "We connected to peer {} for unknown reasons.", peer_id),
-                                Added => info!(target: "validator-network", "New outgoing connection to peer {}.", peer_id),
-                                Replaced => info!(target: "validator-network", "Replaced outgoing connection to peer {}.", peer_id),
-                            },
-                            None => self.spawn_new_outgoing(peer_id, addresses, outgoing_result_for_parent.clone()),
+                Some((peer_id, maybe_data_for_network, connection_type)) = worker_results.next() => {
+                    use ConnectionType::*;
+                    let spawn_new_legacy_connection = match connection_type {
+                        LegacyIncoming => {
+                            self.manager.remove_peer(&peer_id);
```

Comment:
Shouldn't all peer removal be based on whether `maybe_data_for_network.is_some()`? This would simplify a lot and seems like good logic.

Reply:
No, because the spawning of new threads depends on this status, and that is in the `None` branch. :/ I moved some of the logic around and added some comments, perhaps now it's more readable. I also think I fixed a (minor but embarrassing) bug on the way.